services-core/src/gas_price/gnosis_safe.rs

```diff
@@ -0,0 +1,157 @@
+    match network_id {
+        "1" => Some(DEFAULT_MAINNET_URI),
+        "4" => Some(DEFAULT_RINKEBY_URI),
+        _ => None,
+    }
+}
+
+/// Gas prices in wei retrieved from the gas station. This is a result from the
+/// API call.
+#[derive(Deserialize, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct GasPrices {
+    pub last_update: String,
+    #[serde(with = "display_fromstr")]
+    pub lowest: f64,
+    #[serde(with = "display_fromstr")]
+    pub safe_low: f64,
+    #[serde(with = "display_fromstr")]
+    pub standard: f64,
+    #[serde(with = "display_fromstr")]
+    pub fast: f64,
+    #[serde(with = "display_fromstr")]
+    pub fastest: f64,
+}
+
+const FAST_PERCENTILE: f64 = 0.75;
+const STANDARD_PERCENTILE: f64 = 0.5;
+const SAFE_LOW_PERCENTILE: f64 = 0.3;
+
+const SECONDS_PER_BLOCK: f64 = 15.0;
+// Treat percentiles as probabilities for geometric distribution.
+const FAST_TIME: f64 = SECONDS_PER_BLOCK * 1.0 / FAST_PERCENTILE;
+const STANDARD_TIME: f64 = SECONDS_PER_BLOCK * 1.0 / STANDARD_PERCENTILE;
+const SAFE_LOW_TIME: f64 = SECONDS_PER_BLOCK * 1.0 / SAFE_LOW_PERCENTILE;
+
+/// Retrieve gas prices from the Gnosis Safe gas station service.
+#[derive(Debug)]
+pub struct GnosisSafeGasStation {
+    client: HttpClient,
+    uri: Uri,
+}
+
+impl GnosisSafeGasStation {
+    pub fn new(http_factory: &HttpFactory, api_uri: &str) -> Result<GnosisSafeGasStation> {
+        let client = http_factory.create()?;
+        let uri: Uri = api_uri.parse()?;
+        Ok(GnosisSafeGasStation { client, uri })
+    }
+
+    /// Retrieves the current gas prices from the gas station.
+    pub async fn gas_prices(&self) -> Result<GasPrices> {
+        self.client
+            .get_json_async(&self.uri, HttpLabel::GasStation)
+            .await
+    }
+}
+
+#[async_trait::async_trait]
+impl GasPriceEstimating for GnosisSafeGasStation {
+    // The default implementation calls estimate_with_limits with 30 seconds which would result in
+    // the standard time instead of fast. So to keep that behavior we implement it manually.
+    async fn estimate(&self) -> Result<U256> {
+        let response = self.gas_prices().await?;
+        Ok(num::f64_to_u256(response.fast))
+    }
+
+    async fn estimate_with_limits(&self, gas_limit: U256, time_limit: Duration) -> Result<U256> {
+        let response = self.gas_prices().await?;
+        let result = estimate_with_limits(&response, gas_limit, time_limit)?;
+        Ok(num::f64_to_u256(result))
+    }
+}
+
+fn estimate_with_limits(
+    response: &GasPrices,
+    _gas_limit: U256,
+    time_limit: Duration,
+) -> Result<f64> {
+    // We insert two extra points for the linear interpolation because this gas estimator reacts
+    // slowly to gas price changes which means that for example in times of rising gas prices
+    // fast might not be fast enough.
+    let points: &[(f64, f64)] = &[
+        (0.0, response.fast * 2.0),
```

Comment:
Isn't there also `fastest` and `lowest`?

Reply:
Explained in the thread below.